### DISCUSSION TITLE
[FEATURE]

### DISCUSSION THREAD

**Discussion by daliagachc** [bokeh/bokeh#12695] [Internals and design (Q&A)]
### Problem description

i often do exploratory analysis with bokeh in many multiple notebooks and not always save the output. 

### Feature description

i wish there was a way to programatically extract this plots into either static images of interactive htmls so that i can review them 

### Potential alternatives

there are options for doing this with matplotlib. see for example: 
https://github.com/damienmarlier51/Junix

**Comment by bryevdv** [comment]
Up front: moving this to a dev discussion, because I think there is realistically not any chance that this would be developed in the main bokeh repo. Just as with the MPL tool, this would be better as a separately developed project. If that happened and there was interest and commitment to long-term maintenance, we could perhaps discuss including under the Bokeh GH org in the future. 

I should also describe at a very high level what a solution might look like for Bokeh, and also why it is substantially more complicated than the MPL case. In the MPL case, actual base64-encoded images are stored as string data in the notebook output cells. It only takes a trivial few lines of code to un-unencode those strings and write them to disk as-is.

The Bokeh case is not like that, at all. Bokeh output is a set of instruction that have to be executed by a JavaScript runtime. There is no way to generate Bokeh images without running JS code through a browser JS engine. Bokeh can generate PNGs by driving a headless browser (chrome or firefox) using seleniumm, and, more or less, taking a screenshot. Any analogous tool for pulling Bokeh content out of notebooks would have to do the same. Extract the BokehJS JSON output, feed it to a headless browser, to generate a plot in the headless browser, and then capture an image. That's more work, but more importantly, a lot more dependencies to manage. 

Also up front: I can't speak for any other core devs in @bokeh/core but I am not personally interested in spending any time pursuing this directly, but I'm happy to try to offer technical guidance to anyone who does.

**Reply by daliagachc** [reply]
Thank you for the heads up. I do understand now that it is more involved than the MPL alternative. Just in case somebody gets interested too, I leave some discussion I found on the jupyter discussion board that points in a possible direction: 
https://discourse.jupyter.org/t/how-to-extract-bokeh-figure-without-coding-inside-ipynb/11810

"Hi,

Im currently trying to save all figures from notebooks without opening them into lab/hub in order to get a “thumbnail” of my notebooks.

For matplotlib figure, it is pretty straight forward :
i use :

rst_exporter = RSTExporter() # 
(_, resources) = rst_exporter.from_notebook_node(myNb)
But for holoview/bokeh, it is a bit more tricky
i tried to use HTMLExporter with selecting the figure cell. But then nothing appears in the html file
So i also included the cell with the holoview import like this :


 html_exporter_with_figs = HTMLExporter()
 html_exporter_with_figs.preprocessors

 nb = nbformat.read("ExempleBokeh.ipynb", as_version=4)
 nb.cells=nb.cells[1:2]+nb.cells[6:8]
 #print(nb.cells[0])
 (s, resources_with_fig) = html_exporter_with_figs.from_notebook_node(nb)
 #print(resources_with_fig)
 with open("test2.html","w") as file:
     file.write(str(s))
it is working, i got my holoview figure correctly displayed. But i also got all the cell decoration, code … etc … like a classic notebook. By hand, i can remove all the “div” i dont want, but i think there is a better solution.

What i would like is ONLY the figure :
Is there example somewhere of preprocessor or postprocessor to “convert” into html div only the output ?

was not easy to find but options in Exporter did the trick : c = Config() c.HTMLExporter.exclude_input_prompt =True c.HTMLExporter.exclude_output_prompt =True c.HTMLExporter.exclude_input =True # create the new exporter using the custom config html_exporter_with_figs = HTMLExporter(config = c) [htt…](https://nbconvert.readthedocs.io/en/latest/config_options.html#exporter-options)
"
